R/PythonPkgWrapperUtils.R

```diff
@@ -0,0 +1,443 @@
+# ------------------------------------------------------------------------------
+#
+#   Helpers for wrapping python packages
+#
+# ------------------------------------------------------------------------------
+
+#' Determines args and kwargs
+#'
+#' This function determines args and kwargs for a given list of values.
+#'
+#' @param ... the input values
+#' @return A list of args and kwargs
+determineArgsAndKwArgs <- function(...) {
+  values <- list(...)
+  valuenames <- names(values)
+  n <- length(values)
+  args <- list()
+  kwargs <- list()
+  if (n > 0) {
+    positionalArgument <- TRUE
+    for (i in 1:n) {
+      if (is.null(valuenames) || length(valuenames[[i]]) == 0 || nchar(valuenames[[i]]) == 0) {
+        # it's a positional argument
+        if (!positionalArgument) {
+          stop("positional argument follows keyword argument")
+        }
+        if (is.null(values[[i]])) {
+          # inserting a value into a list at best is a no-op, at worst removes an existing value
+          # to get the desired insertion we must wrap it in a list
+          args[length(args) + 1] <- list(NULL)
+        } else {
+          args[[length(args) + 1]] <- values[[i]]
+        }
+      } else {
+        # It's a keyword argument.  All subsequent arguments must also be keyword arg's
+        positionalArgument <- FALSE
+        # a repeated value will overwite an earlier one
+        if (is.null(values[[i]])) {
+          # inserting a value into a list at best is a no-op, at worst removes an existing value
+          # to get the desired insertion we must wrap it in a list
+          kwargs[valuenames[[i]]] <- list(NULL)
+        } else {
+          kwargs[[valuenames[[i]]]] <- values[[i]]
+        }
+      }
+    }
+  }
+  list(args = args, kwargs = kwargs)
+}
+
+#' Clean up stack trace
```

Comment:
The purpose of this function is to remove the Python stack trace from an error message generated when calling Python from R.  This makes the command line response more readable when an error occurs.  To support debugging the stack trace truncation can be overridden by setting the global option 'verbose' to TRUE.